File: python-playwright/src/utilities/logger.py
# Created on : Mar 28, 2026, 4:10:12 PM
# Author     : Tahreem J. Naseem (www.tahreems.com / +92-333-443-8775) {TECHBYTES}


import logging
from logging.handlers import RotatingFileHandler
import os
from pathlib import Path
import re

from src.configurations.configuration_loader import ConfigurationLoader


def _to_log_level(level_str: str | None, default=logging.INFO) -> int:
    if not level_str:
        return default
    return getattr(logging, str(level_str).upper(), default)


def _get_worker_id() -> str:
    return os.getenv("PYTEST_XDIST_WORKER", "main")


def _sanitize_name(name: str) -> str:
    return "".join(c if c.isalnum() or c in ("-", "_", ".") else "_" for c in name)


def _build_formatter() -> logging.Formatter:
    return logging.Formatter(
        "%(asctime)s | %(levelname)-8s | %(processName)s | %(threadName)s | %(name)s | %(message)s"
    )


def _parse_max_bytes(value, default: int = 5 * 1024 * 1024) -> int:
    if isinstance(value, int):
        return value

    if isinstance(value, str):
        sanitized = value.replace(" ", "")
        if re.fullmatch(r"\d+(?:\*\d+)*", sanitized):
            result = 1
            for chunk in sanitized.split("*"):
                result *= int(chunk)
            return result

        if sanitized.isdigit():
            return int(sanitized)

    return default


def get_logger(name: str) -> logging.Logger:
    """
    Shared logger for framework/services/pages/modules.

    Uses one log file per worker:
      logs/<env>/<worker>.log
    """
    logger = logging.getLogger(name)

    if not logger.handlers:
        file_level = _to_log_level(ConfigurationLoader.get("log_level", "INFO"))
        console_level = _to_log_level(ConfigurationLoader.get("console_log_level", "INFO"))
        log_dir = Path(str(ConfigurationLoader.get("log_dir", "logs/dev")))
        worker_id = _get_worker_id()
        logger.setLevel(logging.INFO)
        
        log_dir.mkdir(parents=True, exist_ok=True)
        log_file = log_dir / f"{worker_id}.log"

        logger.setLevel(min(file_level, console_level))
        # logger.propagate = False
        logger.propagate = True

        # console_handler = logging.StreamHandler()
        formatter = _build_formatter()
        # console_handler.setLevel(console_level)
        # console_handler.setFormatter(formatter)
        file_handler = RotatingFileHandler(
            filename = log_file,
            maxBytes = _parse_max_bytes(ConfigurationLoader.get("max_file_size", 5 * 1024 * 1024)),
            backupCount = int(ConfigurationLoader.get("max_old_files_num") or 5),
            encoding = ConfigurationLoader.get("log_encoding", "utf-8"),
        )
        file_handler.setLevel(file_level)
        file_handler.setFormatter(formatter)
        # logger.addHandler(console_handler)
        logger.addHandler(file_handler)

    return logger


def get_test_logger(nodeid: str) -> logging.Logger:
    """
    Test-specific logger.
    Logger name is based on the pytest nodeid so logs are easier to trace.

    Example:
      tests/test_login.py::test_valid_login
    """
    safe_name = _sanitize_name(nodeid)
    return get_logger(f"test.{safe_name}")

File: javascript-playwright/src/utilities/logger.js
import fs from "node:fs";
import path from "node:path";
import { getConfigValue } from "../configurations/configuration-loader.js";
import { ROOT_DIR } from "./path-loader.js";

const levels = { DEBUG: 10, INFO: 20, WARN: 30, ERROR: 40 };

function parseMaxBytes(value, defaultValue = 5 * 1024 * 1024) {
  if (Number.isInteger(value)) return value;
  if (typeof value === "string" && /^\d+(?:\s*\*\s*\d+)*$/.test(value)) {
    return value.split("*").map((chunk) => Number(chunk.trim())).reduce((a, b) => a * b, 1);
  }
  return defaultValue;
}

function shouldLog(configured, level) {
  return levels[level] >= levels[String(configured || "INFO").toUpperCase()];
}

function sanitizeName(name) {
  return name.replace(/[^a-zA-Z0-9_.-]/g, "_");
}

export function getLogger(name) {
  const logLevel = getConfigValue("log_level", "INFO");
  const consoleLevel = getConfigValue("console_log_level", "INFO");
  const logDir = path.join(ROOT_DIR, getConfigValue("log_dir", "logs/dev"));
  fs.mkdirSync(logDir, { recursive: true });
  const worker = process.env.TEST_WORKER_INDEX ? `worker-${process.env.TEST_WORKER_INDEX}` : "main";
  const logFile = path.join(logDir, `${worker}.log`);
  const maxBytes = parseMaxBytes(getConfigValue("max_file_size", 5 * 1024 * 1024));

  function write(level, message) {
    const line = `${new Date().toISOString()} | ${level.padEnd(8)} | ${name} | ${message}\n`;
    if (shouldLog(logLevel, level)) {
      if (fs.existsSync(logFile) && fs.statSync(logFile).size > maxBytes) {
        fs.renameSync(logFile, `${logFile}.${Date.now()}.bak`);
      }
      fs.appendFileSync(logFile, line, getConfigValue("log_encoding", "utf-8"));
    }
    if (shouldLog(consoleLevel, level)) {
      process.stdout.write(line);
    }
  }

  return {
    debug: (message) => write("DEBUG", message),
    info: (message) => write("INFO", message),
    warn: (message) => write("WARN", message),
    error: (message) => write("ERROR", message)
  };
}

export function getTestLogger(testId) {
  return getLogger(`test.${sanitizeName(testId)}`);
}
